src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ExceptionsShouldBePublic.cs

```diff
@@ -24,16 +26,50 @@ public abstract class ExceptionsShouldBePublicAnalyzer : DiagnosticAnalyzer
                                                                              s_localizableMessage,
                                                                              DiagnosticCategory.Design,
                                                                              DiagnosticSeverity.Warning,
-                                                                             isEnabledByDefault: false,
+                                                                             isEnabledByDefault: true,
                                                                              description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: "https://msdn.microsoft.com/en-us/library/bb264484.aspx",
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
         public override void Initialize(AnalysisContext analysisContext)
         {
-            
+            analysisContext.RegisterSymbolAction(AnalyzeException, SymbolKind.NamedType);
+        }
+
+        private void AnalyzeException(SymbolAnalysisContext context)
+        {            
+            var symbol = (INamedTypeSymbol)context.Symbol;
+
+            // skip public symbols
+            if (symbol.IsPublic()) return;
+
+            // only report if base type matches 
+            var baseType = symbol.BaseType;
+            if (IsExceptionType(context, baseType))
+            {                
+                context.ReportDiagnostic(symbol.CreateDiagnostic(Rule));
+            }
+        }
+
+        private bool IsExceptionType(SymbolAnalysisContext context, INamedTypeSymbol baseType)
+        {
+            return GetExceptionTypes(context).Contains(baseType);
+        }
+
+        private HashSet<INamedTypeSymbol> exceptionTypes;
```

Comment:
Don't do this. Since an analyzer lives for the life time of VS, this will field will hold on to the exception symbols from the very first compilation and that will cause the entire compilation to stay alive causing a leak. Worse, for new compilation the symbol for System.Exception might not be the same as the one for the first compilation and this could result in a functional error as well.

Register a CompilationStartAction and put this HashSet there so that it goes away when the compilation is done.